backend/sync.js

```diff
@@ -379,22 +379,21 @@ function generateSyncMessage(backend, syncState) {
     return [syncState, null]
   }
 
-  if (sentHashes.length > 0 && changesToSend.length > 0) {
-    let sentHashesMap = {}
-    for (const hash of sentHashes) {
-      sentHashesMap[hash] = true
-    }
-    changesToSend = changesToSend.filter(change => !sentHashesMap[decodeChangeMeta(change, true).hash])
+  if (Object.keys(sentHashes).length > 0 && changesToSend.length > 0) {
+    changesToSend = changesToSend.filter(change => !sentHashes[decodeChangeMeta(change, true).hash])
   }
 
   // Regular response to a sync message: send any changes that the other node
   // doesn't have. We leave the "have" field empty because the previous message
   // generated by `syncStart` already indicated what changes we have.
   const syncMessage = {heads: ourHeads, have: ourHave, need: ourNeed, changes: changesToSend}
-  const hashesToSend = changesToSend.map(change => decodeChangeMeta(change, true).hash)
+  for (const change of changesToSend) {
+    sentHashes[decodeChangeMeta(change, true).hash] = true
+  }
```

Comment:
Note that this mutates the `sentHashes` object that is passed in. In previous conversations, @pvh was adamant that these functions should be totally side-effect-free.

Reply:
Fixed in 6468bafb01bbe313f831bce5ce789be878cc5459